### DISCUSSION TITLE
Autocompletion using expression parameter sub-types (layer, color...)

### DISCUSSION THREAD

**Discussion by D8H** [4ian/GDevelop#2678] [Enhancement ideas (and their technical discussions)]
## Use case
Expressions like MouseX(:Layer, :number) have a general string autocompletion for its layer parameter. It could be interesting to refine autocompletion using the subtype "Layer" instead of the type "string".
https://trello.com/c/hdJPfWtC/200-add-dropdown-with-list-of-points-for-parameters-asking-for-a-sprite-point-name-in-events

## Code limitations
The GUI calls gd.ExpressionCompletionFinder.getCompletionDescriptionsFor() which only gives string or number as type for parameters no matter the subtype.
https://github.com/4ian/GDevelop/blob/98303f7f50b22ebb1928d3282c7c093fdaa891ec/newIDE/app/src/EventsSheet/ParameterFields/GenericExpressionField/index.js#L340-L343

Looking at the Core code, the parser loses the information on the subtype when it builds the tree:
https://github.com/4ian/GDevelop/blob/98303f7f50b22ebb1928d3282c7c093fdaa891ec/Core/GDCore/Events/Parsers/ExpressionParser2.h#L579-L586

The subtype to type mapping is in IsExpression()
https://github.com/4ian/GDevelop/blob/98303f7f50b22ebb1928d3282c7c093fdaa891ec/Core/GDCore/Extensions/Metadata/ParameterMetadata.h#L174-L196

## Solution
How should the subtype make its way to the GUI?
I guess that adding it to ExpressionNode may not be ideal as it would only be useful for instances that are the parameter child of a FunctionCallNode.
But when a user is writing a parameter, getCompletionDescriptionsFor will give the description of the ExpressionNode that represents the parameter.

**Comment by D8H** [comment]
I think it could be done by modifying ExpressionCompletionFinder to:
* Get the parent of nodeAtLocation from ExpressionNodeLocationFinder
* Make a specific behavior in OnVisitNumberNode, OnVisitTextNode and OnVisitVariableNode if the parent is a FunctionCallNode.

https://github.com/4ian/GDevelop/blob/822146e5f11a32612a757c0174f8a9caba205fa8/Core/GDCore/IDE/Events/ExpressionCompletionFinder.h#L178-L197

**Reply by 4ian** [reply]
> Make a specific behavior in OnVisitNumberNode, OnVisitTextNode and OnVisitVariableNode if the parent is a FunctionCallNode.

I'm a bit worried of this kind of approach, intuitively it sounds a bit hacky (but maybe I've not entirely understood the issue).

**Reply by D8H** [reply]
At first, I though that passing the refined type to the children could be a solution.
But, for instance, `"Layer"` in `MouseX("Layer", 0)` is a layer only because of its parent MouseX. So, to me, it feels more logical to use the parent rather than a side effect of the parent on its child.

And in case of autocompletion, having the refined node type is not enough because if the expression is `MouseX("Layer" + Variable(layerIndex), 0)` and the cursor inside `"Layer"`, I think the autocompletion should stay generic.

**Reply by 4ian** [reply]
> But, for instance, "Layer" in MouseX("Layer", 0) is a layer only because of its parent MouseX

Sure, but this is not forbidding us to pass to the children expression node "your a string, but more precisely you represents a layer" right?

The parser will create a representation like this:
```
FunctionCallNode with 2 parameters:
- TextNode (which is a ExpressionNode with type "string")
- NumberNode (which is a ExpressionNode with type "number")
```

if we pass more information, we might be able to create completions for TextNode that are the possible values for the layers.

> having the refined node type is not enough because if the expression is MouseX("Layer" + Variable(layerIndex), 0) and the cursor inside "Layer", I think the autocompletion should stay generic.

We could surely display the additional completions (here, the layer names) only if the expression is composed of a single text node or empty?

**Reply by D8H** [reply]
> Sure, but this is not forbidding us to pass to the children expression node "your a string, but more precisely you represents a layer" right?

Do you think it should affect only the 1st depth child or propagate until another function call?

> We could surely display the additional completions (here, the layer names) only if the expression is composed of a single text node or empty?

Yes, I think the parent is needed to write this condition. This is not really an issue because it can be get without much calculus:
* Adding `parentNode = foundNode;` before `foundNode = &node;`
https://github.com/4ian/GDevelop/blob/516b1d978c5883287ba99ee545d7d11f1b07b8eb/Core/GDCore/IDE/Events/ExpressionNodeLocationFinder.h#L108-L118

**Comment by 4ian** [comment]
> I guess that adding it to ExpressionNode may not be ideal as it would only be useful for instances that are the parameter child of a FunctionCallNode.

It could be maybe still useful to make the "subtype" survives the parsing. Could maybe be used for something else.
So it might be worth checking this approach (passing the "original, concrete type" to the nodes).

**Comment by D8H** [comment]
I did an implementation here: https://github.com/4ian/GDevelop/pull/2702

**Reply by 4ian** [reply]
Looks nice, I think using the expressionMetadata is the way to go! 👍 Sounds fairly clean and adapted.
